src/lib/statistics/chunk_statistics/histograms/abstract_histogram.hpp

```diff
@@ -0,0 +1,203 @@
+   * The list is sorted by distinct value from lowest to highest.
+   */
+  static std::vector<std::pair<T, uint64_t>> _calculate_value_counts(const std::shared_ptr<const BaseSegment>& segment);
+
+  /**
+   * Given a map of distinct values to their respective number of occurrences,
+   * return a list sorted by distinct value from lowest to highest and return it.
+   */
+  static std::vector<std::pair<T, uint64_t>> _sort_value_counts(const std::unordered_map<T, uint64_t>& value_counts);
+
+  /**
+   * Calculates the estimated cardinality for predicate types supported by all data types.
+   */
+  float _estimate_cardinality(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                              const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Makes pruning decisions for predicate types supported by all data types.
+   */
+  bool _can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                  const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Given a value return its numerical representation.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  uint64_t _convert_string_to_number_representation(const std::string& value) const;
+
+  /**
+   * Given a numerical representation of a string return the string.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  std::string _convert_number_representation_to_string(const uint64_t value) const;
+
+  /**
+   * Returns the share of values in a bin that are lower than `value`.
+   * This method is specialized for strings.
+   */
+  float _bin_share(const BinID bin_id, const T value) const;
+
+  /**
+   * Returns the width of a bin.
+   * This method is specialized for strings to return a numerical width.
+   */
+  virtual HistogramWidthType _bin_width(const BinID index) const;
+
+  /**
+   * Returns the id of the bin that holds the given `value`.
+   * Returns INVALID_BIN_ID if `value` does not belong to any bin.
+   */
+  virtual BinID _bin_for_value(const T value) const = 0;
+
+  /**
+   * Returns the id of the bin after the one that holds the given `value`.
+   * If `value` does not belong to any bin but is smaller than max(), it is in a gap.
+   * In that case return the bin right after the gap.
+   * If the bin that holds the value is the last bin or it is greater than max, return INVALID_BIN_ID.
+   */
+  virtual BinID _upper_bound_for_value(const T value) const = 0;
+
+  /**
+   * Returns the smallest value in the bin.
+   */
+  virtual T _bin_min(const BinID index) const = 0;
+
+  /**
+   * Returns the largest value in a bin.
+   */
+  virtual T _bin_max(const BinID index) const = 0;
+
+  /**
+   * Returns the number of values in a bin.
+   */
+  virtual uint64_t _bin_count(const BinID index) const = 0;
```

Comment:
`bin_height`? otherwise I might think this returns the number of bins or something.